File: server/backend/internal/wifi/wifi.go
// Package wifi provides the same wireless abstraction Luna uses: wpa_cli on
// headless systems, with a no-op provider when no wireless interface exists.
package wifi

import (
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

// Network is a visible Wi-Fi network.
type Network struct {
	SSID      string `json:"ssid"`
	Signal    int    `json:"signal"`
	Encrypted bool   `json:"encrypted"`
}

// Status describes the current Wi-Fi association.
type Status struct {
	Available bool   `json:"available"`
	Connected bool   `json:"connected"`
	SSID      string `json:"ssid,omitempty"`
	IPAddress string `json:"ip_address,omitempty"`
	State     string `json:"state"`
}

// Provider is the wireless backend. Implementations are process-safe and
// never log passphrases.
type Provider interface {
	Scan() ([]Network, error)
	Connect(ssid, passphrase string) error
	Status() (Status, error)
	Forget() error
}

// WpaCli drives wpa_supplicant through wpa_cli (headless/Alpine path).
type WpaCli struct {
	Interface string
}

// New returns a WpaCli provider bound to the named interface.
func New(iface string) Provider {
	return &WpaCli{Interface: iface}
}

// Auto returns a WpaCli provider bound to the first wireless interface, or a
// Noop provider when the device has no wireless radio at all. This is the
// production entry point so callers don't have to sniff interfaces themselves.
func Auto() Provider {
	if iface := FindWirelessInterface(); iface != "" {
		return &WpaCli{Interface: iface}
	}
	return Noop{}
}

// Noop reports Wi-Fi unavailable.
type Noop struct{}

func (Noop) Scan() ([]Network, error)     { return nil, fmt.Errorf("Wi-Fi isn't available on this device") }
func (Noop) Connect(string, string) error { return fmt.Errorf("Wi-Fi isn't available on this device") }
func (Noop) Forget() error                { return fmt.Errorf("Wi-Fi isn't available on this device") }
func (Noop) Status() (Status, error) {
	return Status{Available: false, State: "unavailable"}, nil
}

func (w *WpaCli) run(args ...string) (string, error) {
	full := append([]string{"-i", w.Interface}, args...)
	cmd := exec.Command("wpa_cli", full...)
	out, err := cmd.Output()
	if err != nil {
		if exit, ok := err.(*exec.ExitError); ok {
			return "", fmt.Errorf("the Wi-Fi tool said: %s", strings.TrimSpace(string(exit.Stderr)))
		}
		return "", fmt.Errorf("could not reach the Wi-Fi tool: %w", err)
	}
	return string(out), nil
}

func (w *WpaCli) Scan() ([]Network, error) {
	if _, err := w.run("scan"); err != nil {
		return nil, err
	}
	time.Sleep(1500 * time.Millisecond)
	out, err := w.run("scan_results")
	if err != nil {
		return nil, err
	}
	return parseScanResults(out), nil
}

func (w *WpaCli) Connect(ssid, passphrase string) error {
	id, err := w.run("add_network")
	if err != nil {
		return err
	}
	id = strings.TrimSpace(id)
	if _, err := w.run("set_network", id, "ssid", fmt.Sprintf("%q", ssid)); err != nil {
		return err
	}
	if passphrase == "" {
		if _, err := w.run("set_network", id, "key_mgmt", "NONE"); err != nil {
			return err
		}
	} else {
		if _, err := w.run("set_network", id, "psk", fmt.Sprintf("%q", passphrase)); err != nil {
			return err
		}
	}
	if _, err := w.run("enable_network", id); err != nil {
		return err
	}
	if _, err := w.run("select_network", id); err != nil {
		return err
	}
	_, _ = w.run("save_config")
	time.Sleep(3000 * time.Millisecond)
	status, err := w.Status()
	if err != nil {
		return err
	}
	if !status.Connected {
		return fmt.Errorf("That password didn't work. Check the sticker on your internet box and try again.")
	}
	return nil
}

func (w *WpaCli) Status() (Status, error) {
	out, err := w.run("status")
	if err != nil {
		return Status{Available: true, State: "error"}, err
	}
	status := Status{Available: true}
	for _, line := range strings.Split(out, "\n") {
		key, value, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		switch key {
		case "ssid":
			status.SSID = value
		case "ip_address":
			status.IPAddress = value
		case "wpa_state":
			status.State = value
		}
	}
	status.Connected = status.State == "COMPLETED"
	return status, nil
}

func (w *WpaCli) Forget() error {
	out, err := w.run("status")
	if err == nil {
		for _, line := range strings.Split(out, "\n") {
			key, value, ok := strings.Cut(line, "=")
			if ok && key == "id" {
				_, _ = w.run("remove_network", value)
				_, _ = w.run("save_config")
				return nil
			}
		}
	}
	return nil
}

func parseScanResults(out string) []Network {
	var networks []Network
	lines := strings.Split(out, "\n")
	for _, line := range lines[1:] {
		cols := strings.Split(line, "\t")
		if len(cols) < 5 {
			continue
		}
		ssid := strings.TrimSpace(cols[4])
		if ssid == "" {
			continue
		}
		signal, _ := strconv.Atoi(strings.TrimSpace(cols[2]))
		flags := cols[3]
		encrypted := strings.Contains(flags, "PSK") || strings.Contains(flags, "WPA") || strings.Contains(flags, "WEP")
		networks = append(networks, Network{SSID: ssid, Signal: signal, Encrypted: encrypted})
	}
	return networks
}

// FindWirelessInterface returns the first /sys/class/net interface with a
// wireless subdirectory, or "".
func FindWirelessInterface() string {
	entries, err := os.ReadDir("/sys/class/net")
	if err != nil {
		return ""
	}
	for _, entry := range entries {
		if _, err := os.Stat("/sys/class/net/" + entry.Name() + "/wireless"); err == nil {
			return entry.Name()
		}
	}
	return ""
}

// EthernetConnected reports whether any non-wireless interface has carrier.
func EthernetConnected() bool {
	entries, err := os.ReadDir("/sys/class/net")
	if err != nil {
		return false
	}
	for _, entry := range entries {
		name := entry.Name()
		if name == "lo" {
			continue
		}
		if _, err := os.Stat("/sys/class/net/" + name + "/wireless"); err == nil {
			continue
		}
		if carrier, err := os.ReadFile("/sys/class/net/" + name + "/carrier"); err == nil && strings.TrimSpace(string(carrier)) == "1" {
			return true
		}
	}
	return false
}

File: luna/crates/lunad/src/wifi.rs
//! Wi-Fi provider abstraction.
//!
//! Luna OS uses wpa_supplicant through `wpa_cli` (no dbus dependency).
//! LibreServ can later add a NetworkManager provider behind the same trait.
//! Passphrases are passed to wpa_supplicant and never logged or returned.

use std::process::Command;
use std::sync::Mutex;

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
pub struct WifiNetwork {
    pub ssid: String,
    pub signal: i32,
    pub encrypted: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
pub struct WifiStatus {
    pub available: bool,
    pub connected: bool,
    pub ssid: Option<String>,
    pub ip_address: Option<String>,
    pub state: String,
}

#[derive(Debug, thiserror::Error)]
pub enum WifiError {
    #[error("Wi-Fi isn't available on this device.")]
    Unavailable,
    #[error("That password didn't work. Check the sticker on your internet box and try again.")]
    Auth,
    #[error("That network name or password has characters Luna can't use. Try another.")]
    InvalidInput,
    #[error("Could not reach the Wi-Fi tool: {0}")]
    Io(#[source] std::io::Error),
    #[error("The Wi-Fi tool said: {0}")]
    Tool(String),
}

pub trait WifiProvider: Send + Sync {
    fn scan(&self) -> Result<Vec<WifiNetwork>, WifiError>;
    fn connect(&self, ssid: &str, passphrase: &str) -> Result<(), WifiError>;
    fn status(&self) -> Result<WifiStatus, WifiError>;
    fn forget(&self) -> Result<(), WifiError>;
}

/// Provider for headless systems using `wpa_cli`.
pub struct WpaCliProvider {
    interface: String,
}

impl WpaCliProvider {
    pub fn new(interface: impl Into<String>) -> Self {
        Self {
            interface: interface.into(),
        }
    }

    fn run(&self, args: &[&str]) -> Result<String, WifiError> {
        let out = Command::new("wpa_cli")
            .arg("-i")
            .arg(&self.interface)
            .args(args)
            .output()
            .map_err(WifiError::Io)?;
        if !out.status.success() {
            return Err(WifiError::Tool(
                String::from_utf8_lossy(&out.stderr).trim().to_string(),
            ));
        }
        Ok(String::from_utf8_lossy(&out.stdout).into_owned())
    }
}

impl WifiProvider for WpaCliProvider {
    fn scan(&self) -> Result<Vec<WifiNetwork>, WifiError> {
        self.run(&["scan"])?;
        // wpa_supplicant reports results asynchronously; a short settle is required.
        std::thread::sleep(std::time::Duration::from_millis(1500));
        let output = self.run(&["scan_results"])?;
        Ok(parse_scan_results(&output))
    }

    fn connect(&self, ssid: &str, passphrase: &str) -> Result<(), WifiError> {
        let ssid_q = wpa_cli_quoted(ssid)?;
        let id = self.run(&["add_network"])?.trim().to_string();
        self.run(&["set_network", &id, "ssid", &ssid_q])?;
        if passphrase.is_empty() {
            self.run(&["set_network", &id, "key_mgmt", "NONE"])?;
        } else {
            let psk_q = wpa_cli_quoted(passphrase)?;
            self.run(&["set_network", &id, "psk", &psk_q])?;
        }
        self.run(&["enable_network", &id])?;
        self.run(&["select_network", &id])?;
        let _ = self.run(&["save_config"]);

        // Verify the association actually completed.
        std::thread::sleep(std::time::Duration::from_millis(3000));
        let status = self.status()?;
        if status.connected {
            Ok(())
        } else {
            Err(WifiError::Auth)
        }
    }

    fn status(&self) -> Result<WifiStatus, WifiError> {
        let output = self.run(&["status"])?;
        let mut ssid = None;
        let mut ip = None;
        let mut state = "unknown".to_string();
        for line in output.lines() {
            let Some((key, value)) = line.split_once('=') else {
                continue;
            };
            match key {
                "ssid" => ssid = Some(value.to_string()),
                "ip_address" => ip = Some(value.to_string()),
                "wpa_state" => state = value.to_string(),
                _ => {}
            }
        }
        let connected = state == "COMPLETED";
        Ok(WifiStatus {
            available: true,
            connected,
            ssid,
            ip_address: ip,
            state,
        })
    }

    fn forget(&self) -> Result<(), WifiError> {
        let output = self.run(&["status"])?;
        for line in output.lines() {
            if let Some((key, value)) = line.split_once('=')
                && key == "id"
            {
                self.run(&["remove_network", value])?;
                let _ = self.run(&["save_config"]);
                return Ok(());
            }
        }
        Ok(())
    }
}

/// Used when no wireless interface or wpa_cli exists.
pub struct NoopProvider;

impl WifiProvider for NoopProvider {
    fn scan(&self) -> Result<Vec<WifiNetwork>, WifiError> {
        Err(WifiError::Unavailable)
    }
    fn connect(&self, _ssid: &str, _passphrase: &str) -> Result<(), WifiError> {
        Err(WifiError::Unavailable)
    }
    fn status(&self) -> Result<WifiStatus, WifiError> {
        Ok(WifiStatus {
            available: false,
            connected: false,
            ssid: None,
            ip_address: None,
            state: "unavailable".into(),
        })
    }
    fn forget(&self) -> Result<(), WifiError> {
        Err(WifiError::Unavailable)
    }
}

/// Deterministic provider for tests.
#[derive(Default)]
pub struct MockProvider {
    pub networks: Mutex<Vec<WifiNetwork>>,
    pub connected_ssid: Mutex<Option<String>>,
    pub connect_calls: Mutex<Vec<(String, String)>>,
}

impl WifiProvider for MockProvider {
    fn scan(&self) -> Result<Vec<WifiNetwork>, WifiError> {
        Ok(self.networks.lock().unwrap().clone())
    }
    fn connect(&self, ssid: &str, passphrase: &str) -> Result<(), WifiError> {
        self.connect_calls
            .lock()
            .unwrap()
            .push((ssid.to_string(), passphrase.to_string()));
        *self.connected_ssid.lock().unwrap() = Some(ssid.to_string());
        Ok(())
    }
    fn status(&self) -> Result<WifiStatus, WifiError> {
        let ssid = self.connected_ssid.lock().unwrap().clone();
        Ok(WifiStatus {
            available: true,
            connected: ssid.is_some(),
            ssid: ssid.clone(),
            ip_address: Some("192.168.1.42".into()),
            state: if ssid.is_some() {
                "COMPLETED".into()
            } else {
                "DISCONNECTED".into()
            },
        })
    }
    fn forget(&self) -> Result<(), WifiError> {
        *self.connected_ssid.lock().unwrap() = None;
        Ok(())
    }
}

/// Quote a value for `wpa_cli set_network`. Reject quotes, backslashes, and
/// newlines so they cannot break out of the quoted argument.
pub(crate) fn wpa_cli_quoted(value: &str) -> Result<String, WifiError> {
    if value
        .chars()
        .any(|c| c == '"' || c == '\\' || c == '\n' || c == '\r' || c == '\0')
    {
        return Err(WifiError::InvalidInput);
    }
    Ok(format!("\"{value}\""))
}

fn parse_scan_results(output: &str) -> Vec<WifiNetwork> {
    let mut networks = Vec::new();
    for line in output.lines().skip(1) {
        let cols: Vec<&str> = line.split('\t').collect();
        if cols.len() < 5 {
            continue;
        }
        let ssid = cols[4].trim();
        if ssid.is_empty() {
            continue;
        }
        let signal = cols[2].trim().parse().unwrap_or(0);
        let flags = cols[3];
        let encrypted = flags.contains("PSK") || flags.contains("WPA") || flags.contains("WEP");
        networks.push(WifiNetwork {
            ssid: ssid.to_string(),
            signal,
            encrypted,
        });
    }
    networks.sort_by_key(|n| std::cmp::Reverse(n.signal));
    networks
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_wpa_cli_scan_results() {
        let output = "bssid / frequency / signal level / flags / ssid\n\
                      00:11:22:33:44:55\t2412\t-44\t[WPA2-PSK-CCMP][ESS]\tHome WiFi\n\
                      aa:bb:cc:dd:ee:ff\t2412\t-80\t[ESS]\tCoffee Shop\n";
        let nets = parse_scan_results(output);
        assert_eq!(nets.len(), 2);
        assert_eq!(nets[0].ssid, "Home WiFi");
        assert!(nets[0].encrypted);
        assert_eq!(nets[1].ssid, "Coffee Shop");
        assert!(!nets[1].encrypted);
    }

    #[test]
    fn mock_provider_round_trip() {
        let mock = MockProvider::default();
        mock.networks.lock().unwrap().push(WifiNetwork {
            ssid: "Home".into(),
            signal: -42,
            encrypted: true,
        });
        assert_eq!(mock.scan().unwrap().len(), 1);
        mock.connect("Home", "hunter2").unwrap();
        assert!(mock.status().unwrap().connected);
        mock.forget().unwrap();
        assert!(!mock.status().unwrap().connected);
    }

    #[test]
    fn wpa_cli_quoted_rejects_breakouts() {
        assert_eq!(wpa_cli_quoted("Home WiFi").unwrap(), "\"Home WiFi\"");
        assert!(wpa_cli_quoted("x\"\nset_network").is_err());
        assert!(wpa_cli_quoted("a\\b").is_err());
        assert!(wpa_cli_quoted("line\n2").is_err());
    }
}
